compiler/damlc/lib/DA/Cli/Damlc.hs

```diff
@@ -942,6 +952,69 @@ getMultiPackagePath multiPackageLocation =
+commaSepAnd _ = []
+
+-- Gets a list of the (transitive) sub-project directories that define given composite dar name 
+findCompositeDarTransitiveDefiners :: MultiPackageConfigFields -> LF.PackageName -> [FilePath]
+findCompositeDarTransitiveDefiners mpc darName =
+  mapMaybe (\(path, darNames) -> if darName `elem` darNames then Just path else Nothing) $ mpTransitiveCompositeDarNames mpc
+
+-- Prints a message warning if a composite dar exists in one or more sub-projects
+warnTransitiveCompositeDar :: Handle -> MultiPackageConfigFields -> LF.PackageName -> IO ()
+warnTransitiveCompositeDar handle mpc name = do
+  let otherDefiners = findCompositeDarTransitiveDefiners mpc name
+   in when (not $ null otherDefiners) $ do
+        hPutStrLn handle $ "Warning: Found definition for " <> T.unpack (LF.unPackageName name)
+          <> " in the following sub-projects:"
+        traverse_ (hPutStrLn handle . ("  - " <>)) otherDefiners
+        hPutStrLn handle $ "Use `--multi-package-path` to specify the multi-package.yaml "
+          <> "defining the composite-dar if you intended to build one of these composite dars."
+
+data MultiPackageBuildModeComposite a = SelectCompositeDars [a] | AllCompositeDars
+
+-- a is either CompositeDar or LF.PackageName
+data MultiPackageBuildMode a
+  = SinglePackage PackageConfigFields
+  | MultiPackage MultiPackageBuildAll (MultiPackageBuildModeComposite a)
+
+toMaybePackageConfig :: MultiPackageBuildMode a -> Maybe PackageConfigFields
+toMaybePackageConfig (SinglePackage pkgConfig) = Just pkgConfig
+toMaybePackageConfig _ = Nothing
+
+realiseMultiBuildMode :: MultiPackageConfigFields -> MultiPackageBuildMode LF.PackageName -> IO (MultiPackageBuildMode CompositeDar)
+realiseMultiBuildMode multiPackageConfig buildMode =
+  case buildMode of
+    SinglePackage pkgConfig -> pure $ SinglePackage pkgConfig
+    MultiPackage buildAll (SelectCompositeDars names) -> MultiPackage buildAll . SelectCompositeDars <$> extractCompositeDars names
+    MultiPackage buildAll AllCompositeDars -> pure $ MultiPackage buildAll AllCompositeDars
+  where
+    extractCompositeDars :: [LF.PackageName] -> IO [CompositeDar]
+    extractCompositeDars darNames = do
+      compositeDars <-
+        forM darNames $ \darName -> do
+          let isMatchingCompositeDar cd =
+                cdName cd == darName ||
+                  (LF.unPackageName (cdName cd) <> "-" <> LF.unPackageVersion (cdVersion cd)) == LF.unPackageName darName
+              matchingCompositeDars = filter isMatchingCompositeDar $ mpCompositeDars multiPackageConfig
+          case matchingCompositeDars of
+            [dar] -> pure dar
+            [] -> do
+              hPutStrLn stderr $ "Couldn't find composite dar with the name " <> T.unpack (LF.unPackageName darName) <> " in " <> mpPath multiPackageConfig
+              warnTransitiveCompositeDar stderr multiPackageConfig darName
+              exitFailure
```

Comment:
Can we exit with failure after `forM` is done running, s.t. an error/warning is printed for every composite dar asked for? This maximizes the amount of useful info for a call.